editor/plugins/camera_3d_editor_plugin.cpp

```diff
@@ -79,37 +82,69 @@ Camera3DEditor::Camera3DEditor() {
 	preview->connect("pressed", callable_mp(this, &Camera3DEditor::_pressed));
 }
 
+static Size2i _get_viewport_size(Camera3D *p_camera) {
+	Viewport *viewport = p_camera->get_viewport();
+
+	Window *window = Object::cast_to<Window>(viewport);
+	if (window) {
+		return window->get_size();
+	}
+
+	SubViewport *sub_viewport = Object::cast_to<SubViewport>(viewport);
+	ERR_FAIL_NULL_V(sub_viewport, Size2i());
+
+	if (sub_viewport == EditorNode::get_singleton()->get_scene_root()) {
+		return Size2(GLOBAL_GET("display/window/size/viewport_width"), GLOBAL_GET("display/window/size/viewport_height"));
+	}
+
+	return sub_viewport->get_size();
+}
```

Comment:
Are there really this many cases to determine the size of a Camera's viewport? Seems like some logic that should be in a more central place.

This probably should be some helper function on Camera3D class itself. I see duplicate code in `editor/plugins/gizmos/camera_3d_gizmo_plugin.cpp` and I wouldn't be surprised if it were also present in other places, such as Camera2D.

Reply:
Yep, the code should be deduplicated.
It can be in Node3DEditor singleton. Camera3D is ok too, but it's editor-only code, so better not.